host/sgx/linux/vdso.c

```diff
@@ -0,0 +1,217 @@
+#include "../exception.h"
+#include "sgx.h" // Linux kernel header
+
+extern bool oe_is_avx_enabled;
+
+static elf64_sym_t _sgx_enter_enclave_sym;
+static vdso_sgx_enter_enclave_t _vdso_sgx_enter_enclave;
+
+oe_result_t oe_sgx_initialize_vdso(void)
+{
+    oe_result_t result = OE_NOT_FOUND;
+    void* sgx_vdso_base = (void*)getauxval(AT_SYSINFO_EHDR);
+
+    if (!sgx_vdso_base)
+        goto done;
+
+    if (elf64_find_dynamic_symbol_by_name_with_header(
+            (const elf64_ehdr_t*)sgx_vdso_base,
+            "__vdso_sgx_enter_enclave",
+            &_sgx_enter_enclave_sym) != 0)
+        goto done;
+
+    _vdso_sgx_enter_enclave = (vdso_sgx_enter_enclave_t)(
+        (uint64_t)sgx_vdso_base + _sgx_enter_enclave_sym.st_value);
+
+    result = OE_OK;
+
+done:
+    if (result == OE_OK)
+        OE_TRACE_INFO("vDSO symbols found. Opt into oe_vdso_enter.");
+    else
+        OE_TRACE_INFO("vDSO symbols not found. Fallback to regular oe_enter "
+                      "implementation.");
+
+    return result;
+}
+
+typedef struct _vdso_args
+{
+    uint64_t rdi;
+    uint64_t rsi;
+} vdso_args_t;
+
+static int oe_vdso_user_handler(
+    long rdi,
+    long rsi,
+    long rdx,
+    long rsp,
+    long r8,
+    long r9,
+    struct sgx_enclave_run* run)
+{
+    uint64_t arg1 = (uint64_t)rdi;
+    uint64_t arg2 = (uint64_t)rsi;
+    vdso_args_t* return_args = NULL;
+    int result = 0;
+
+    OE_UNUSED(rdx);
+    OE_UNUSED(rsp);
+    OE_UNUSED(r8);
+    OE_UNUSED(r9);
+
+    if (!run)
+    {
+        result = -1;
+        goto done;
+    }
+
+    return_args = (vdso_args_t*)run->user_data;
+
+    switch (run->function)
+    {
+        case ENCLU_EENTER:
+            /* Unexpected case (e.g., the enclave loses EPC context
+             * because of power events). Return failing value. */
+            result = -1;
+            break;
+        case ENCLU_EEXIT:
+        {
+            /* Regular exit (the enclave finishes an ECALL or makes an
+             * OCALL). Return zero.
+             * Note that an alternative implementation is returning
+             * ENCLU_EENTER. However, doing so requires setting up
+             * the input parameters into corresponding registers (e.g.,
+             * rdi, rsi, and rdx) and ensuring the compiler to preserve
+             * these registers until the function returns. Instead,
+             * we return zero to avoid dealing with such complexities
+             * and also to use similar implementation as regular enter. */
+            return_args->rdi = arg1;
+            return_args->rsi = arg2;
+            result = 0;
+            break;
+        }
+        case ENCLU_ERESUME:
+        {
+            /* Hardware exceptions occur */
+
+            oe_host_exception_context_t host_context = {0};
+
+            host_context.rax = ENCLU_ERESUME;
+            host_context.rbx = run->tcs;
+
+            /* AEP is assigned by vDSO implementation */
+
+            oe_host_handle_exception(&host_context);
```

Comment:
Given that exceptions are handled in this manner, we no longer need to register a host signal handler.